visual-java/src/main/java/com/saucelabs/visual/VisualApi.java

```diff
@@ -617,36 +618,53 @@ public <D> D execute(GraphQLOperation operation, Class<D> responseType) {
   }
 
   private String sauceVisualCheckLocal(String snapshotName, CheckOptions options) {
+    Capabilities caps = driver.getCapabilities();
     Window window = new Window(this.driver);
-    Rectangle viewport = window.getViewport();
-
-    byte[] screenshot;
-
-    // clip image if required
+    Map<String, Object> windowDims =
+        (Map<String, Object>)
+            driver.executeScript(
+                "return { height: window.innerHeight, width: window.innerWidth, dpr: window.devicePixelRatio }");
+    double devicePixelRatio = formatDevicePixelRatio(windowDims.get("dpr"));
+    boolean fullPage = options.getFullPageScreenshotConfig() != null;
     WebElement clipElement = getClipElement(options);
+    Point scrollOffset = fullPage ? new Point(0, 0) : window.getViewport().getPoint();
```

Comment:
Does it make sense to set `scrollOffset` to `window.getViewport().getPoint()` if `clipElement == null` (i.e. as an `else` statement after `if (clipElement != null)`)? I think it is always overriden if `clipElement` is not `null`.

Reply:
I don't think that'd be necessary since the default is set here (and we'd just be overriding it in the else). 

And that's correct on clip front -- at least for non full page screenshots. We will always attempt to scroll the element into view so we update the position here after that call.

Reply:
If I'm reading this correctly, we're reading it twice needlessly. Assume that `clipElement` is not `null` and `fullPage` is `false`:
1. `scrollOffset` is set to `window.getViewport().getPoint()` on line `630`,
2. `scrollOffset` is set to `window.getViewport().getPoint()` on line `638` again, and it's not used before at all.

If `clipElement` would be `null` and `fullPage` is `false`:
1. `scrollOffset` is set to `window.getViewport().getPoint()` on line `630`.

The latter case could be set in `else` to the condition on `line `634`, and it would then fix double retrieval from the former case. Note that `window.getViewport()` makes a call to Selenium.

Reply:
Updated the behavior here and moved the scrollOffset to a single check following the clipElement block